pkg/core/native/name_service.go

```diff
@@ -0,0 +1,727 @@
+package native
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"regexp"
+	"sort"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/dao"
+	"github.com/nspcc-dev/neo-go/pkg/core/interop"
+	"github.com/nspcc-dev/neo-go/pkg/core/interop/runtime"
+	"github.com/nspcc-dev/neo-go/pkg/core/native/nativenames"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/core/storage"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/encoding/bigint"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/callflag"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/manifest"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+)
+
+// NameService represents native NameService contract.
+type NameService struct {
+	nonfungible
+	NEO *NEO
+}
+
+type nameState struct {
+	state.NFTTokenState
+	// Expiration is token expiration height.
+	Expiration uint64
+	// HasAdmin is true if token has admin.
+	HasAdmin bool
+	// Admin is token admin.
+	Admin util.Uint160
+}
+
+// RecordType represents name record type.
+type RecordType byte
+
+// Pre-defined record types.
+const (
+	RecordTypeA     RecordType = 1
+	RecordTypeCNAME RecordType = 5
+	RecordTypeTXT   RecordType = 16
+	RecordTypeAAA   RecordType = 28
+)
+
+const (
+	nameServiceID = -7
+
+	prefixRoots       = 10
+	prefixDomainPrice = 22
+	prefixExpiration  = 20
+	prefixRecord      = 12
+
+	secondsInYear = 365 * 24 * 3600
+
+	// DefaultDomainPrice is the default price of register method.
+	DefaultDomainPrice = 10_00000000
+	// MinDomainNameLength is minimum domain length.
+	MinDomainNameLength = 3
+	// MaxDomainNameLength is maximum domain length.
+	MaxDomainNameLength = 255
+)
+
+var rootRegex, nameRegex, ipv4Regex, ipv6Regex *regexp.Regexp
+
+func init() {
+	var err error
+	rootRegex, err = regexp.Compile("^[a-z][a-z0-9]{0,15}$")
+	if err != nil {
+		panic(err)
+	}
+	// Lookahead is not supported by Go, but it is simple `(?=.{3,255}$)`,
+	// so we check name length explicitly.
+	nameRegex, err = regexp.Compile("^([a-z0-9]{1,62}\\.)+[a-z][a-z0-9]{0,15}$")
+	if err != nil {
+		panic(err)
+	}
+	ipv4Regex, err = regexp.Compile("^(2(5[0-5]|[0-4]\\d))|1?\\d{1,2}(\\.((2(5[0-5]|[0-4]\\d))|1?\\d{1,2})){3}$")
```

Comment:
`net.ParseIP`?

Reply:
I am not sure about corner cases, that's why it is here which Go can allow or disallow, that's why here is a copy.

Reply:
OK, let's also keep this.